### ISSUE TITLE
Dropout 적용에 관한 질문

### ISSUE THREAD

**Issue by hccho2** [NLP-kr/tensorflow-ml-nlp#19]
지난 주말에 책 내용을 공부하다 보니, 질문이 좀 많아졌습니다.

Appendex - Seq2Seq 챗봇

https://github.com/NLP-kr/tensorflow-ml-nlp/blob/master/6.CHATBOT/Appendix/model.py

Dropout에 관해 2가지 질문드립니다.

1. 어텐션 부분에서 
Line 166: 
input_token_emb = tf.concat([context_vector, input_token_emb], axis=-1)
input_token_emb = tf.keras.layers.Dropout(0.5)(input_token_emb)

이 부분에서, contex_vector와 input_token_emb를 concat한 후, dropout을 적용했습니다.

이 부분은 다음과 같이 하려는 의도가 아니었는지 
 ==> (input_token_emb -> dropout적용) 후, context_vector와 concat하는 방식

이 질문은 모델 구성에 관한 문제라 최종적인 성능에서의 영향은 있을 수도 있지만, 맞고 틀리고 하는 문제는 아닙니다.

그래서 조심스럽게 질문드립니다.

2. Dropout 이중 적용 문제:
Line: 170에서
decoder_outputs, decoder_state = rnn_cell(input_token_emb, decoder_state)
decoder_outputs = tf.keras.layers.Dropout(0.5)(decoder_outputs)

이 코드에서 rnn_cell은 이미 DropoutWrapper를 거치고 나왔는데, 다시 Dropout을 적용하고 있습니다.

이 부분은 혹시 의도하지 않은 것은 아닌지 궁금합니다.

**Comment by changwookjun** [comment]
안녕하세요 . 전창욱 입니다.

1. 현재 소스는 여러 테스트 실험을 통해 최대 성능 지표를 가지는 소스 구조를 가지고 있습니다. 
의도한 부분이며 Operation 동작 이후에 Dropout을 주기 위해 만들어 졌습니다. 
말씀하신것처럼 다양한 방법으로 구현 가능해 보입니다.  ^^

2. cell에 대한 Dropout과 Layer에 대한 Dropout으로 나뉘는것 같습니다. 
rnn_cell 의 Dropout같은 경우에는 input, forget, output gate의 input과 output을 통제하여 Cell state의 영향을 주는 Dropout으로 이해 했습니다. 
tensorflow.org : https://www.tensorflow.org/api_docs/python/tf/nn/rnn_cell/DropoutWrapper
source : https://github.com/tensorflow/tensorflow/blob/456fbc0e498e3d10604973de9f46ca48d62267cc/tensorflow/python/ops/rnn_cell_impl.py#L1203
paper : https://arxiv.org/abs/1512.05287

tf.keras.layers.Dropout의 Dropout같은 경우에는 Layer의 입력에 Dropout을 적용하는 부분으로 이해 했습니다.
tensorflow.org : https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dropout
source : https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/layers/core.py#L111-L175

그래서 두개의 Dropout은 다른 부분 적용인 듯 합니다.  ^^

좋은 질문 감사드립니다.

**Comment by hccho2** [comment]
여기 구현 소스를 보면

https://github.com/tensorflow/tensorflow/blob/456fbc0e498e3d10604973de9f46ca48d62267cc/tensorflow/python/ops/rnn_cell_impl.py#L1433




tf.nn.rnn_cell.DropoutWrapper(cell,input_keep_prob,  output_keep_prob, state_keep_prob)

이렇게 했을 때,

inputs와 state가 있으면,

```
inputs = DropOut(intputs,input_keep_prob)

outputs, new_state = cell(inputs,state)
new_state = DropOut(new_state,state_keep_prob)

output = DropOut(outputs,output_keep_prob)

return output,new_state
```

이런 구조입니다.

교재의 모델에서는 input_keep_prob, state_keep_prob은 주지않고, output_keep_prob만 주었습니다.

그래서 최종적인 outputs가 Dropout을 한번 거치고 나온 결과이고, 다음 단계에서 다시 Dropout을 적용하고 있으니, 이중적용이 아닌가 생각됩니다.

**Comment by changwookjun** [comment]
그럴수도 있겠네요 . 

만약 이중 적용이 된다면 어떻게 될까요 ?  

궁금하긴 하네요 .. ~~^^

좋은 지적 감사드립니다.

**Comment by hccho2** [comment]
이중 적용된다고 모델에 심각한 하자가 생기는 것은 아닐 것이고, 의도한 keep rate이 제곱되어 적용되는 정도 이겠지요~~

답변 감사합니다.

**Comment by changwookjun** [comment]
그럴수 있을듯 합니다. ~~^^
